### DISCUSSION TITLE
Shapes for Products and Services

### DISCUSSION THREAD

**Discussion by jeff-zucker** [solid-contrib/practitioners#7] [Ideas]
A space to discuss  [the proposed shape for products and services](https://github.com/solid-contrib/practitioners/tree/main/working-groups/shape-for-products-and-services).

**Comment by jeff-zucker** [comment]
Some other terms that might be useful in some circumstances : 

 schema:countryOfOrigin                                                                           
 schema:funding                                                                                   
 schema:rating                                                                                    
 schema:review

**Reply by danielbakas** [reply]
Good ones. Many of the CreativeWork, Product, SoftwareApplication, Organization and Project might be useful too.

Maybe finding specific examples of Products and Services will allow us to collect attributes to make these abstract classes more robust. Although I'm worried that the shapes (being so abstract) could possibly lead to too many attributes. I imagine you're focusing on those which most Products and Services share (?)

**Reply by prozion** [reply]
> Maybe finding specific examples of Products and Services will allow us to collect attributes to make these abstract classes more robust.

Hello! Recently I had been engaged in my hobby project to collect RDF of individuals, that are Solid applications or libraries. It was a part of broader collection of "solid objects".

Currently I am on the trip, away from my laptop, so excuse me for not generating standalone RDF with products/services, but in the *.tree source code it looks like that: https://codeberg.org/prozion/solid-kgr/src/branch/main/source/facts/software.tree

In addition to the property `lang` (programming language), I applied properties `contributor` and its subproperty `developer`. Also I started to mark what frameworks and libraries have been used in the product's source code (following my selfish interest to track those, who develop in Clojure programming language, just like me :-) )

**Comment by danielbakas** [comment]
Hi Jeff! This is great.

For a very long time, at Semantyk, We have defined product and service, both as children of “Project”.

This has allowed us to define basic attributes that both products and services share.

This is also also allowed us to manage composition (“organization has project”).

Could it be useful for us to first define a basic, shared set of attributes for both, which could in turn contribute to interoperability between products and services?

Thanks again for opening this space. It’s really valuable 🤗

**Comment by jeff-zucker** [comment]
We want to be able to find out how recently the description for a product or service was updated and by whom.  That is a feature of the description, not of the product.  It is not, for example the same as the release date of a product and so should not be a property of the product.  So I suggest we describe it something like this
```turtle
<> # this docuemnt                                                                           
  foaf:primaryTopic :MyProduct ;                                                         
  schema:dateModified "" ;; # the date the description was changed                      
  schema:maintainer <> .     # webid or email of person who last updated the description     
                                                                                             
:MyProduct # the product this document describes                                           
  a goodrelations:ProductOrService ;                                                                  
  schema:releaseDate "" ;   # the date the product was changed 
  # ... 
```

[Edited, on elf's suggestion, I removed solid:ProductOrServiceDescription -  there's no need for a separate type for that]

**Reply by danielbakas** [reply]
So the document would describe the Product?

**Reply by jeff-zucker** [reply]
Yes. The same way a WebID Profile describes a person or agent.   We update the profile document, not the person.

**Reply by prozion** [reply]
Interesting, in this solution - how to keep track of all updates of description, not just the last one?

**Comment by jeff-zucker** [comment]
# Using GoodRelations and the Product Ontology 

The GoodRelations (gr:) ontology is one of, if not the, most widely adopted e-commerce ontologies and has a broad selection of ways to describe products and services. Therefore, I believe we should treat the class `gr:ProductOrService` as fundamental.  

In the documentation for gr: we find : "Your first choice for specializations of gr:ProductOrService should be http://www.productontology.org." The Product Ontology (pto:) is a collaboration between Good Relations, Schema.org, and Wikipedia which turns almost any Wikipedia web page into an ontology term for a product or service.  For example `pto:Authentication_server` and `pto:Tutorial` are subclasses of `gr:ProductOrService`.

Now, why should we use `pto:Authentication_server` instead of just a string like "Authentication Server"?  Well interoperability is one answer.  Another is localization.  Since pto:Authentication_server is defined on Wikipedia, it includes multiple language versions of it's label, for example `rdfs:label "Authentifizierungsserver"@de;` and `rdfs:label "驗證伺服器"@zh;`.  And if your language isn't represented there, Wikipedia has a way for you to contribute it.  So `pto:` is very extendable and currently has over 300,000 products and services.

This all sounds great.  Here's one caveat - GoodRelations is e-**commerce**.  It is about commerce.  It has  `gr:BusinessFunction` but no `:SocialFunction`.  Schema has terms for buyer and seller but not for volunteer. We should keep these gaps in mind and try to fill some of them. The pto: ontology should come in handy there.

Here are the relevant prefixes : 


@prefix gr: <https://www.heppnetz.de/ontologies/goodrelations/v1.html#&gt; .
@prefix pto: <http://www.productontology.org/id/&gt; .

With the pto: you can add .ttl to see turtle rather than HTML, e.g. http://www.productontology.org/id/Authentication_server.ttl .

To find any given term, just look it up on wikipedia.  If wikipedia has a page for it, copy the last part of the page name e.g. Authtentication_server and add the pto: prefix to it to get the term.

**Reply by danielbakas** [reply]
I love this. This is a great approach. Could we maybe expand on GoodRelations/PTO to include this SocialFunction?

This is related to your question on social benefit projects on the linked data channels right @jeff-zucker? And the issues of missing definitions for these.

Now, if we're thinking in terms of Wikipedia pages as classes or things, why not use Wikidata or DBPedia instead? 😯

**Reply by danielbakas** [reply]
Also, would it be too soon to think in terms of "App" or "Platform" as a subclass of Product, Service or ProductOrService? I can think of a lot of reasons for the Practitioners and Onboarding groups to define at least this one very popular and useful subclass. Or would it be a different parent class altogether?

Curious to know what you think

**Reply by jeff-zucker** [reply]
> Could we maybe expand on GoodRelations/PTO to include this SocialFunction?

I am keeping a list of Classes and Predicates that relate  to nonprofits.  The chances are that GoodRelations would not be interested since they specify e-commerce as their domain.  A more likely place is schema.org.  I spoke to timbl about this and he will set up meetings with schema.org when we have a better list. We shouldn't rush this.  Let's keep track of predicates and classes we find missing and then make a proposal to schema.org.

> why not use Wikidata or DBPedia instead?

Because neither of them has a productontology.org and becaue they are both  based on Wikipedia and because there is a three-way collaboration between Schema.org, Wikipedia, and GoodRelations.

> would it be too soon to think in terms of "App" or "Platform" as a subclass

Here is my preliminary list of top-level subclasses of gr:ProductOrService relevant to the Solid Ecosystem  : 
```turtle
      sioc:Forum  # chats, forums, email lists, ...                                     
      pto:Meeting  # meetings, events, symposia, ...                                     
      schema:SoftwareApplication  # apps                                                                
      pto:Authentication_server # identity servers/services                                           
      pto:Storage_server # storage servers/services                                            
      pto:Programming_tool # developer tools                                                     
      pto:Ontology_information_science  # # ontologies                                                          
      pto:Specification_technical_standard # specifications                                                      
      pto:Web_resource  # web sites/services, documents, media, ...                           
      schema:Action #  actions and activites not listed above 
```

**Reply by danielbakas** [reply]
Excellent! Thanks @jeff-zucker!
